Add bootstrap command

ale8k opened the pull request:
## Description
Implements bootstrap cmd for JIMM.

## Engineering checklist
<!-- *Check only items that apply* -->

- [ ] Documentation updated
- [x] Covered by unit tests
- [ ] Covered by integration tests
- [x] Covered by QA test 

## Test instructions
Q/A via:
Set lxd to be available on 8443 `lxc config set core.https_address "[::]:8443"`

Run juju add-cloud --client
	Type: lxd
	Name: lxdtest
	Endpoint: https://127.0.0.1:8443 (It should show cert auth is available when set)
	Rest just press enter

```go
// Copyright 2025 Canonical.

package main

import (
	"context"
	"fmt"
	"os"
	"path/filepath"

	jujucloud "github.com/juju/juju/cloud"

	"github.com/canonical/jimm/v3/internal/jujucommands"
)

func main() {
	ctx := context.Background()

	absPath, err := filepath.Abs("./jujudata")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error getting absolute path: %v\n", err)
		os.Exit(1)
	}
	runner := jujucommands.NewCommandRunner(absPath)
	cmd := jujucommands.NewBootstrapCmd(runner)

	cred := jujucloud.NewCredential(jujucloud.CertificateAuthType, map[string]string{
		"server-cert": ``,
		"client-cert": ``,
		"client-key": ``,
	})

	cloudCred := *jujucloud.NewEmptyCloudCredential()
	// This is the default region's credential (when a credential is added, it is always tied
	// to the provided CloudNameAndRegion) implicitly by our command.
	cloudCred.AuthCredentials["default"] = cred

	p := jujucommands.BootstrapCmdParams{
		CloudNameAndRegion:   "testing-lxd",
		ControllerName:       "my-controller",
		LoginTokenRefreshURL: "http://myurl.com",

		PersonalCloud: jujucloud.Cloud{
			Type: "lxd",
			Name: "testing-lxd",
			AuthTypes: jujucloud.AuthTypes{
				jujucloud.CertificateAuthType,
			},
			Endpoint: "https://127.0.0.1:8443",
			Regions: []jujucloud.Region{
				{
					Name:     "default",
					Endpoint: "https://127.0.0.1:8443",
				},
			},
		},
		CloudCred: cloudCred,
	}
	fmt.Println("Starting")
	retriever, store, cleanup, err := cmd.Run(ctx, p)
	if err != nil {
		panic(err)
	}
	_ = cleanup
	_ = store

	for line := range retriever {
		if line.Err != nil {
			fmt.Println(line.Err)
			os.Exit(1)
		}
		fmt.Println(line.Line)
	}
}

```

ale8k committed:
feat: add bootstrap command

ale8k committed:
feat: wip

ale8k committed:
feat: bootstrap cmd implementation

ale8k committed:
feat: enable ssh keys to be loaded

ale8k committed:
feat: return memstore to access controller, accounts, and models

alesstimec commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,224 @@
+	cred jujucloud.CloudCredential,
+	pubKey []byte,
+	privKey []byte,
+) (<-chan outputLine, jujuclient.ClientStore, func(), error) {
```
perhaps we could define a type to be returned by this function that would have these fields..

ale8k replied:
Was discussed in other PR. Going to close, ping if still an issue and I'll do it in a followup.

alesstimec commented on internal/jujucommands/jujucommands.go:
```diff
@@ -0,0 +1,148 @@
+)
+
+var (
+	whitelist = []string{"help", "bootstrap", "update-public-clouds"}
```
we don't need `help`

alesstimec commented on internal/jujucommands/jujucommands.go:
```diff
@@ -0,0 +1,148 @@
+//
+// The OutputLine channel is closed once all output from the command has finished an the error has been captured.
+func runCmdWithOutputRetriever(store jujuclient.ClientStore, cmdAndArgs string) (<-chan outputLine, error) {
+	// memStore := jujuclient.NewEmbeddedMemStore()
```
dd

alesstimec commented on internal/jujucommands/jujucommands.go:
```diff
@@ -0,0 +1,148 @@
+		"",        // unknown param
+		"",        // no help hint
+		whitelist, // whitelist
+		// We set embedded false because some commands can targets either a controller or client.
```
s/targets/target/

alesstimec commented on internal/jujucommands/jujucommands.go:
```diff
@@ -0,0 +1,148 @@
+
+		close(outputCh)
+	}()
+	go runCmd(cmdCtx, cmdWriter, store, cmdFinishedCh, cmdAndArgs)
```
are we passing in cmdWriter just so that we can close it? if so, you can simply close it after L86 above when you know the command has finished.

ale8k replied:
As this was discussed in the other PR, I'm going to resolve it here. Can ping if you still want to check it out?

alesstimec commented on internal/jujucommands/jujucommands_test.go:
```diff
@@ -0,0 +1,73 @@
+type jujucommandsSuite struct{}
+
+func (s *jujucommandsSuite) TestRunCmdWithOutputRetriever(c *qt.C) {
+	outputCh, err := jujucommands.RunCmdWithOutputRetriever(jujuclient.NewEmbeddedMemStore(), "help")
```
whitelisting the `help` commands seems a bit much just for this test.. you can always run `juju update-public-clouds --help` for this test..

ale8k replied:
Again was discussed in other PR, ping if it's still an issue. I don't see it as harmful. I see it as functionally equivalent to having a variable for the sake of a test?

alesstimec commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,224 @@
+	LoginTokenRefreshURL string
+}
+
+func (b BootstrapCmdParams) validate() error {
```
no unit tests for this..

ale8k replied:
On it!

alesstimec commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,224 @@
+	return nil
+}
+
+func (b BootstrapCmdParams) buildBootstrapCmdStr() string {
```
no unit tests for this

ale8k replied:
On it!

alesstimec commented on internal/jujucommands/jujucommands.go:
```diff
@@ -0,0 +1,148 @@
+//	}
+//
+// The OutputLine channel is closed once all output from the command has finished an the error has been captured.
+func runCmdWithOutputRetriever(store jujuclient.ClientStore, cmdAndArgs string) (<-chan outputLine, error) {
```
i would appreciate a unit test for this function.. you could mock the `runCmd` function for the test..

ale8k replied:
This was a part of the other PR. I don't see it as worth it in mocking it when it's very minimal to process a help/--help command. Due to delicacy, I prefer seeing a command legitimately run...

ale8k committed:
chore: add unit tests

ale8k committed:
test: public-cloud.yaml test

ale8k committed:
feat: run commands via os.exec

The function to run commands now uses os.Exec and reads in line exactly like previously. It
additionally uses a shell parser to protect us from bad escapes and an environment override to
protect us from bad environments.

ale8k committed:
chore: rebase to use binary on disk

ale8k committed:
chore: pr comments

ale8k committed:
chore: pr comments

alesstimec commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,218 @@
+	LoginTokenRefreshURL string
+}
+
+func (b BootstrapCmdParams) Validate() error {
```
godoc, please

alesstimec commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,218 @@
+	var builder strings.Builder
+	builder.WriteString("bootstrap")
+
+	builder.WriteString(fmt.Sprintf(" --login-token-refresh-url=%s", b.LoginTokenRefreshURL))
```
```suggestion
	builder.WriteString(fmt.Sprintf("bootstrap --login-token-refresh-url=%s", b.LoginTokenRefreshURL))
```
and remove L61... the `login-token-refresh-url` flag will always be there

alesstimec commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,218 @@
+	}
+
+	// Always add controller name & cloud at the end
+	builder.WriteString(fmt.Sprintf(" %s", b.CloudNameAndRegion))
```
add both in one `WriteString`?

alesstimec commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,218 @@
+
+func (b BootstrapCmdParams) BuildBootstrapCmdStr() string {
+	var builder strings.Builder
+	builder.WriteString("bootstrap")
```
we could use templates instead of `strings.Builder`.. 
```
bootstrap --login-token-refresh-url={{ .loginTokenRefreshURL }} {{ if .agentVersion }} --agent-version={{ .agentVersion }} {{ end }} {{ if.bootstrapTimeout }} --config bootstrap-timeout={{ .bootstrapTimeout }} {{ end }} {{ .cloudNameAndRegion }} {{ .controllerName }}
```

ale8k committed:
chore: update

ale8k committed:
chore: rebase

rebase fixes

ale8k committed:
chore: cleanup

SimoneDutto commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,213 @@
+}
+
+// NewBootstrapCmd creates a new BootstrapCmd with the specified command runner.
+func NewBootstrapCmd(runner Runner) *bootstrapCmd {
```
Adding JIMM here to check clouds, cloud regions, etc are valid and the user has access to it is for another PR?

ale8k replied:
That's for another PR in the job itself, this is just for running the command.

SimoneDutto commented on internal/jujucommands/bootstrap_test.go:
```diff
@@ -0,0 +1,256 @@
+
+func (r *runnerMock) JujuDataDir() string {
+	return r.dataDir
+}
```
we have gomock in the codebase, so you don't need to create this by hand

SimoneDutto commented on internal/jujucommands/bootstrap_test.go:
```diff
@@ -0,0 +1,256 @@
+			"--agent-version=1.1.1",
+			"--config",
+			"bootstrap-timeout=1000",
+			"testregion/testcloud",
+			"my-controller",
+		},
+	)
+
+	p.AgentVersion = ""
+	args = p.BuildBootstrapCmdArgs()
+
+	c.Assert(
+		args,
+		qt.DeepEquals,
+		[]string{
+			"bootstrap",
+			"--config",
+			"login-token-refresh-url=myurl.com",
+			"--config",
+			"bootstrap-timeout=1000",
+			"testregion/testcloud",
+			"my-controller",
+		},
+	)
+
+	p.BootstrapTimeout = 0
+	args = p.BuildBootstrapCmdArgs()
+
+	c.Assert(
+		args,
+		qt.DeepEquals,
+		[]string{
+			"bootstrap",
+			"--config",
+			"login-token-refresh-url=myurl.com",
+			"testregion/testcloud",
+			"my-controller",
+		},
+	)
+}
```
a table test here would be much more readable. In this way is not clear what is the struct that is being validated because you set/unset fields on the flight

kian99 replied:
Additionally, what about just checking that the resulting list of strings doesn't contain config for the field you zeroed out, rather than checking the entire slice?

SimoneDutto commented on internal/jujucommands/bootstrap_test.go:
```diff
@@ -0,0 +1,256 @@
+			return outputCh, nil
+		},
+		dataDir: c.TempDir(),
+	}
```
gomock

SimoneDutto commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,213 @@
+	// And
+	// 		jujucloud.WritePersonalCloudMetadata
+	// TODO: Remove this dependency and write the files manually?
+	if err := os.Setenv("JUJU_DATA", dataDir); err != nil {
```
relying on env variable to be set is not a good idea, what if it's unset by another go-routine, you would never know.
Is there a safer way to do it?

kian99 replied:
It looks like you can use the function `SetJujuXDGDataHome` in `github.com/juju/juju/juju/osenv/home.go` which will set a global variable that is looked up before checking for any env vars. But this will still limit us to 1 bootstrap at a time.

SimoneDutto commented on internal/jujucommands/bootstrap_test.go:
```diff
@@ -0,0 +1,256 @@
+	)
+
+	// This was set to a temp dir until cleanup runs. We can use it to check the file exists.
+	_, err = os.Stat(filepath.Join(os.Getenv("JUJU_DATA"), "clouds.yaml"))
```
relying on an env var in tests can bring to flaky situation because one test is setting it, the other is cleaning it up

ale8k replied:
Ah what I didn't push this. Resolved.

SimoneDutto commented on internal/jujucommands/jujucommands_test.go:
can you add a test where you cancel the context and the command is cancelled? Our jobtracker relies on the fact that it is possible to cancel a job by cancelling the context, but we don't have a test for it

ale8k replied:
Will do.

kian99 commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,213 @@
+// BootstrapCmdParams holds the parameters to bootstrap a controller for JIMM.
+type BootstrapCmdParams struct {
+	// Arguments to be turned into an actual command str.
+	CloudNameAndRegion   string
```
Splitting them into `Cloud` and `Region` would be clearer.

ale8k replied:
We pass cloud/region to the arg... If I split it, I'll have to reconstruct it at some point anyway. I'm happy to split it but not sure if it is worth it?

kian99 replied:
Sure, fine to leave.

kian99 commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,213 @@
+
+	// Additional args required (like adding credential, cloud, etc.) but JIMM will handle.
+
+	// May be left unset, if set, a personal cloud will be created and used for bootstrap.
+	PersonalCloud jujucloud.Cloud
```
What is a personal cloud?

ale8k replied:
I thought the same initially when looking at how Juju bootstraps but tldr;

clouds.yaml -> add-cloud additions (private / personal / user clouds)
public-clouds.yaml -> predefined list of public clouds

Their language is "Personal Cloud" so I went with their language rather than "private" or something similar.

kian99 commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,213 @@
+	}
+	if !isAPublicCloud {
+		// We presume it is a personal cloud
+		// TODO: Check if credential should be cloudname or include region
```
This todo seems like something that should be done for this PR?

ale8k replied:
It is. I need to ask really. Not sure how to test this without my own regions...

ale8k replied:
Checked, it is just name.

kian99 commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,213 @@
+	}
+
+	// Update public clouds
+	// TODO: Make this a command of this package
```
I'm not following what this is trying to say?

ale8k replied:
We have a file for bootstrap command. The running of this command should be a command in our command package too. I.e., bootstrap.go, updatepubliccloud.go. It's not a big deal though I don't think and didn't want to make this PR larger than it needed to be.

kian99 replied:
Ah, I think it's fine to remove the todo then. Or reword it to `move this command to its own file`.

kian99 commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,213 @@
+		}
+	}
+
+	// TODO: check if cloudName should include region, presuming not right now
```
Again, this seems like something to do for this PR?

ale8k replied:
I had a look at other instances where they update the store credential, and they name all vars cloud name. 

And in here: cmd/juju/cloud/updatecredential.go L279

They update credential without region and the region param is just validating:
```go
		if c.Region != "" {
			if err := validCloudRegion(aCloud, c.Region); err != nil {
				logger.Errorf("%v", err)
				ctx.Warningf("Region %q is not valid for cloud %v.", c.Region, cloudName)
				erred = true
				continue
			}
		}
```

But... The commands help text for update-credential is:
```
Before credential is updated, the new content is validated. For some providers, 
cloud credentials are region specific. To validate the credential for a non-default region, 
use --region.
```

Maybe I'm misunderstanding a bit but it kinda looks like it just stores against cloud name. 

But if you look at the credentials file, they're region specific. 

Anyone reading this comment - It'd be good if another pair of eyes could look over this

ale8k replied:
Ahhh nevermind, it is in the credential itself. So it is indeed cloud name.

kian99 commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,213 @@
+	}
+
+	outputRetriever, err := c.runner.RunJujuCmd(ctx, args)
+	return outputRetriever, store, cleanupTmpJujuData, err
```
Can we change this to check the error first and then just return nil at the end. Just a bit clearer.

kian99 commented on internal/jujucommands/bootstrap.go:
```diff
@@ -0,0 +1,213 @@
+// just the cloud name is checked. If a region is specified, but it isn't valid,
+// an error is returned.
+//
+// TODO: Is there a better way to do this? (Rather than look up metadata and loop through)
```
I would just delete this if you haven't found a better way.

kian99 commented on internal/jujucommands/bootstrap_test.go:
```diff
@@ -0,0 +1,256 @@
+	return r.dataDir
+}
+
+func (s *jujucommandsSuite) TestBootstrapCmdParams_Validate(c *qt.C) {
```
Looks like the formatting got messed up in this function. So many new lines

ale8k replied:
Aha, that's me. It read easier, I'll condense it.

kian99 commented on internal/jujucommands/bootstrap_test.go:
```diff
@@ -0,0 +1,256 @@
+	mock := runnerMock{
+		impl: func(ctx context.Context, args []string) (<-chan jujucommands.OutputLine, error) {
+			// Return chan that has one line inside
+			outputCh := make(chan jujucommands.OutputLine, 1)
+			outputCh <- jujucommands.OutputLine{
+				Line: "testing",
+			}
+			close(outputCh)
```
You could simplify this to,

```suggestion
			outputCh := make(chan jujucommands.OutputLine)
			close(outputCh)
```
The same for the other tests.

kian99 commented on internal/jujucommands/jujucommands.go:
```diff
@@ -13,21 +13,46 @@ import (
 )
 
-type outputLine struct {
+var (
+	cmdPrefix = "juju"
+)
```
Since the command runner has a prefix field, you can delete this and then you won't need to do patching from tests.

kian99 commented on internal/jujucommands/jujucommands.go:
```diff
@@ -13,21 +13,46 @@ import (
-	cmd := exec.CommandContext(ctx, cmdPrefix, args...)
-	cmd.Env = append(cmd.Env, "JUJU_DATA="+jujuDataDir)
+func (b *CommandRunner) RunJujuCmd(ctx context.Context, args []string) (<-chan OutputLine, error) {
+	cmd := exec.CommandContext(ctx, b.prefix, args...)
```
Looks like the tests might need an update since this is using the struct's value rather than the global var.

ale8k replied:
The structure was using the global var (which is why it worked) but I've removed it now as you requested. I updated signature of NewCmdRunner to take a prefix.

ale8k committed:
chore: pr comments

ale8k committed:
chore: fix gosec

ale8k committed:
chore: pr comments

kian99 commented on internal/jujucommands/bootstrap_test.go:
```diff
@@ -0,0 +1,248 @@
+	)
+
+	// This was set to a temp dir until cleanup runs. We can use it to check the file exists.
+	_, err = os.Stat(filepath.Join(os.Getenv("JUJU_DATA"), "clouds.yaml"))
```
I think this line needs to be updated now that we are not using an env var anymore

kian99 commented on internal/jujucommands/bootstrap_test.go:
```diff
@@ -0,0 +1,248 @@
+	)
+
+	// Make sure no personal cloud was written.
+	_, err = os.Stat(filepath.Join(os.Getenv("JUJU_DATA"), "clouds.yaml"))
```
This line too

kian99 commented on internal/jujucommands/jujucommands.go:
```diff
@@ -13,21 +13,45 @@ import (
+//
+// dataDir is the JUJU_DATA directory where juju commands will store their data.
+// It must be an ABSOLUTE path.
+func NewCommandRunner(prefix, dataDir string) *CommandRunner {
```
Can we rename `prefix` to `command`?

kian99 commented on internal/jujucommands/jujucommands_test.go:
```diff
@@ -64,8 +80,9 @@ func (s *jujucommandsSuite) TestRunCmdWithOutputRetriever_Error(c *qt.C) {
-	outputCh, err := jujucommands.RunJujuCmd(testCtx, []string{}, "testing-data-is-set")
+
+	runner := jujucommands.NewCommandRunner("env", "testing-data-is-set")
+	outputCh, err := runner.RunJujuCmd(testCtx, []string{})
```
Just for consistency with other files, can you put this at the top please

SimoneDutto reviewed (approved):
be 🐄-tious merging this

ale8k commented:
> be 🐄-tious merging this
@SimoneDutto - I will.
<img width="427" height="640" alt="image" src="https://github.com/user-attachments/assets/3ffd9034-e6f7-45b3-9b36-667da4053b13" />

ale8k committed:
chore: pr comments
